### ISSUE TITLE
Capsule manual resolution

### ISSUE THREAD

**Issue by bojjenclon** [Aweptimum/Strike#13]
In my setup, I define both "dynamic" and "static" colliders - ones that should be moved by collisions and ones that shouldn't. My player is a dynamic Capsule and I have a static Rectangle in the scene. If use I `S.ettle` the two colliders push each other out properly, but if I use `e1:move(mtv.x, mtv.y)` to try and only move the "dynamic" collider and not the static one, the capsule "jumps" around when the circle portions touch the rectangle. Am I doing something wrong or is this an issue with Strike?

I'm using Candran so ignore the invalid Lua syntax.

```lua
-- Check for collisions and resolve
for i = 1, #physics_entities do
  let e1 = physics_entities[i]
  let c1 = e1.collider

  if e1.physics_type == 'static' then
    continue
  end

  for j = 1, #physics_entities do
    if i == j then
      continue
    end

    let e2 = physics_entities[j]
    let c2 = e2.collider

    -- Broad Phase
    if S.aabb(c1, c2) then
      -- Narrow Phase
      let mtv = S.triking(c1, c2)
      if mtv then
        -- S.ettle will push both entities, but for
        -- static collisions we should only move the
        -- first (dynamic) one.

        if e2.physics_type == 'static' then
          e1:move(mtv.x, mtv.y)
        else
          S.ettle(mtv)
        end

        mtv:stow()
      end
    end
  end
end
```

---

Update

After some further testing, it seems there are just some issues with circles and rectangles interacting in general.

**Comment by Aweptimum** [comment]
Hey @bojjenclon, sorry to keep you waiting. Never got a notification :(

What's probably happening is the mtv's reference point is the dynamic body, not the static. A basic way to validate and get the colliders to behave is to use the collider/collided fields in the mtv, since they provide the context for it. Below is an (untested) bit of lua code you can try. Hope the intent is clear

```lua
if e2.physics_type == 'static' then
    -- Reconcile ref with c1/c2
    local ref = mtv.colliderShape:getRoot()
    if ref == c2 then -- `ref` is static c2, push e1 by mtv
        e1:move(mtv.x, mtv.y)
    else
        e1:move(-mtv.x, -mtv.y) -- `ref` is dynamic c1, push e1 by negative mtv
    end
end
```
If that still produces inconsistent behavior, then the problem is most likely with Strike itself.

**Comment by bojjenclon** [comment]
Hey thanks for getting back to me. I changed my logic to:

```lua
      -- Broad Phase
      if S.aabb(c1, c2) then
        -- Narrow Phase
        let mtv = S.triking(c1, c2)
        let owner = mtv and mtv?.collidedShape:getRoot().owner
        if owner == e1 then
          -- S.ettle will push both entities, but for
          -- static collisions we should only move the
          -- first (dynamic) one. For this, we can use
          -- S.hove to just translates the first one.

          if e2.physics_type == 'static' then
            S.hove(mtv)
          else
            S.ettle(mtv)
          end
        end
      end
```
Which I believe is basically equivalent to your example, but I'm still seeing inconsistent behavior. :/

**Comment by Aweptimum** [comment]
I don't believe that's quite the same as it doesn't seem to account for the 2 cases you'll see with one static and one dynamic. The cases being:
1. mtv's `colliderShape` is static (from static's perspective), dynamic needs to be pushed by mtv as is
2. mtv's `colliderShape` is dynamic (from dynamic's perspective), dynamic needs to be pushed negatively

It looks like your code handles case 1, but does not do anything to address case 2
1. If dynamic == e1 == owner == `collidedShape`, then `colliderShape` is e2 (static), and `S.hove` will do what you want
2. If dynamic == e1 != owner == `collidedShape`, then `colliderShape` is e1 (dynamic), and `S.hove` should not be used

Your code doesn't call `S.hove` in the second case, which is good. However, it doesn't look like anything is done to handle this case (at least in the shared snippet). e1 is still penetrating e2, and it needs to be pushed out of e2 by the mtv in the opposite direction the mtv is pointing. If you don't resolve this case, then you'll see phasing. The inconsistency is probably due to the floating point comparisons in `S.triking` that aim to find the minimum penetrating vector. It's possible each frame for the mtv to "swap" reference frames between the capsule's circles and the edge if they generate "exactly" equal mtv's.

I could be totally wrong about your code. And if you disagree with my evaluation, the next step is probably getting a minimum reproducible example to me so I can mess around. I'm less than a week out from recovering from a case of covid, so programming has been really hard the past 3 weeks. Also the reason why I took so long to respond, I wanted to make sure I comprehended your code. Hope you understand.

Regardless, I appreciate you giving Strike a chance! Feel free to find me in the LÖVE discord (sham)
